torch/ao/quantization/fx/_model_report/detector.py

```diff
@@ -0,0 +1,406 @@
+from torch.ao.quantization.fx.graph_module import GraphModule
+from torch.ao.quantization.observer import ObserverBase
+from torch.ao.quantization.qconfig import QConfig
+from torch.nn.qat.modules.conv import _ConvNd as QatConvNd
+from torch.nn.qat.modules.linear import Linear as QatLinear
+
+ABC: Any = ABCMeta(str("ABC"), (object,), {})
+
+# Adding base class for detectors
+class DetectorBase(ABC):
+    r""" Base Detector Module
+    Any detector class should derive from this class.
+
+    Concrete detectors should follow the same general API, which includes:
+    - A method to calculate and return observer insertion points
+        - Should return both the fqns and the Observer class to insert
+    - A method to return a report based on the the detector
+        - Should return a str-based report and dict info in Tuple[str,Dict] format
+    """
+
+    def __init__(self):
+        super(DetectorBase, self).__init__()
+
+    @abstractmethod
+    def determine_observer_insert_points(self, model):
+        r"""
+        Args
+            model (nn.Module or subclass): model to find observer insertion points
+
+        Returns a Tuple of two elements:
+            Set[str] of observer fqns denoting where to insert observers
+            ObserverBase (or subclass): the class (not an instance) of the observer to insert
+        """
+        pass
+
+    @abstractmethod
+    def get_detector_name(self):
+        r""" Returns the name of the current detector """
+        pass
+
+    @abstractmethod
+    def generate_detector_report(self, model):
+        r"""
+        Args
+            model (nn.Module or subclass): model to find observer insertion points
+
+        Returns a Tuple of two elements:
+            Str: string report of the suggested improvements
+            Dict: contains useful data collected by the observer pertinent to this report
+        """
+        pass
+
+class PerChannelDetector(DetectorBase):
+    r""" This class is used to detect if any Linear or Conv layers in a model utilize per_channel quantization.
+        Only Linear and Conv layers can use per_channel as of now so only these two are currently checked.
+
+        per_channel quantization can lead to major benefits in the form of accuracy.
+        Therefore, if the backend used by the user supports it, it is recommended to use
+    """
+
+    # Default map for representing supported per channel quantization modules for different backends
+    DEFAULT_BACKEND_PER_CHANNEL_SUPPORTED_MODULES: Dict[str, Set[Any]] = {
+        "fbgemm": set([nn.Linear, nn.Conv1d, nn.Conv2d, nn.Conv3d, QatLinear, QatConvNd]),
+        "qnnpack": set([nn.Linear, nn.Conv1d, nn.Conv2d, nn.Conv3d, QatLinear, QatConvNd]),
+        "onednn": set([nn.Linear, nn.Conv1d, nn.Conv2d, nn.Conv3d, QatLinear, QatConvNd]),
+    }
+
+    def __init__(self):
+        super(PerChannelDetector, self).__init__()
+
+        # store the backend information
+        self.backend_chosen = torch.backends.quantized.engine
```

Comment:
I'm wondering if this should be an argument, since the backend that user wants to use to in the end (target backend to use in production) may not be the same as `torch.backends.quantized.engine` (current quantized backend selected when user is preparing the model)